alphadia/workflow/optimizer.py

```diff
@@ -0,0 +1,147 @@
+# native imports
+from functools import wraps
+
+import numpy as np
+
+# alpha family imports
+# third party imports
+import pandas as pd
+
+# alphadia imports
+from alphadia.workflow import reporting
+
+
+def check_if_optimum_found(method):
+    """If an optimum has been found, the optimizer should cease to execute its methods to check, initiate and update.
+    This function can be used as a decorator and placed before the methods in an optimizer to prevent execution of the methods if an optimum has been found.
+    If the optimum has been found, the decorator will print the end of optimization message and, for consistency with stop_or_continue methods, return True.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if self.optimal_tolerance is not None:
+            self.end_of_optimization_message()
+            return True
+        return method(self, *args, **kwargs)
+
+    return wrapper
+
+
+class BaseOptimizer:
+    def __init__(
+        self,
+        initial_tolerance: float,
+        reporter: None | reporting.Pipeline | reporting.Backend = None,
+    ):
+        """This class serves as a base class for organizing the search parameter optimization process, which defines the tolerances used for search.
+
+        Parameters
+        ----------
+        initial_tolerance: float
+            The initial tolerance to start the optimization
+        """
+        self.proposed_new_tolerance = initial_tolerance
+        self.tolerances = []
+        self.round = -1
```

Comment:
Could we also look at `len(self.tolerances)`? 

In general, you always wan't to have a piece of information only in one place as there is the risk that it get's out of sync.

Reply:
Do you mean replace self.round with len(self.tolerances)? That seems to make sense to me.